yarn-project/world-state/src/synchronizer/server_world_state_synchronizer.ts

```diff
@@ -129,6 +144,23 @@ export class ServerWorldStateSynchronizer implements WorldStateSynchronizer {
     return this.blockNumber.get() ?? 0;
   }
 
+  private async pause() {
```

Comment:
Instead of these paused/resume semantics, could we just `syncImmediate` and then put a request to `fork` on the job queue?

The job queue is there to give us that kind of atomicity.

Reply:
But can't we run into a situation where the loop in the `runningPromise` runs again (eg while we await for `l2BlockDownloader.pollImmediate`) and enqueues a `collectAndProcessBlocks` **without** a `targetBlock` before we manage do so?

Reply:
Is the issue that we rely on forking at a specific block and the world state can't be allowed to move past that block before we fork? Presumably this is always a problem if we are just slow to call `syncImmediateAndFork`?

If that's the case then yes, we need to stop the main loop performing it's own sync process like you say.

Reply:
Yep, exactly. I expect this issue to go away once we implement the forkable world-state, but for now we need to stop the main sync.